### PR TITLE
Add initial functional tests

### PR DESCRIPTION
**Pull request by dkarlovi**
As suggested in https://github.com/rollerworks/search/issues/179

### CODE REVIEW

**Comment by dkarlovi** [comment]
@sstok is this something you had in mind?

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/.idea/
```
Comment (sstok) [inline .gitignore]: Please remove this, you can use the global .gitignore instead 👍

Comment (dkarlovi) [reply .gitignore]: Done, thanks for the link.

**Comment by sstok** [comment]
Thank you so much for your work so far 👍

**Comment by dkarlovi** [comment]
@sstok can you elaborate a bit on the `registerField()`? How is that supposed to work exactly?

File: tests/Functional/FunctionalElasticsearchTestCase.php [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        return $build ? $fieldSet->getFieldSet('invoice') : $fieldSet;
+    }
+
+    /**
+     * @param QueryConditionGenerator $conditionGenerator
+     */
+    protected function configureConditionGenerator(QueryConditionGenerator $conditionGenerator)
+    {
+        // TODO: mapping field set alias to Elasticsearch property
+        return;
```
Comment (dkarlovi) [inline tests/Functional/FunctionalElasticsearchTestCase.php]: @sstok this is where I'm hoping you can shed some light. :)

File: tests/Functional/FunctionalElasticsearchTestCase.php [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    /**
+     * @param SearchCondition $condition
+     * @param int[]           $expectedIds
+     */
+    protected function assertDocumentsAreFound(SearchCondition $condition, array $expectedIds)
+    {
+        $conditionGenerator = new QueryConditionGenerator($condition);
+        $this->configureConditionGenerator($conditionGenerator);
+
+        // TODO: where do I get these from?
```
Comment (dkarlovi) [inline tests/Functional/FunctionalElasticsearchTestCase.php]: @sstok also here, don't quite understand how you imagined for it to work.

**Comment by sstok** [comment]
The `registerField` method works similar to the Doctrine DBAL/ORM implementation 😃
You should be able to map a search field to one or more fields, but ElasticSearch actually is much more advanced in this topic as (IIRC) you can create an index with combining fields.

It's a while since I worked on the ElasticSearch implementation 😅

**Comment by dkarlovi** [comment]
@sstok sure, we can work on advanced mappings later, for now let's try and get **anything** working. :)

For example, how did you imagine we map `id: 1` to `/invoices/invoices`, property `_id` with the `registerField()`?

File: .php_cs [hunk 4]
```diff
@@ -15,7 +15,7 @@ return PhpCsFixer\Config::create()
         '@Symfony:risky' => true,
         '@PHP70Migration' => true,
         '@PHP71Migration' => true,
-        'array_syntax' => array('syntax' => 'short'),
+        'array_syntax' => ['syntax' => 'short'],
```
Comment (dkarlovi) [inline .php_cs]: This one is ironic. :sob:

**Comment by dkarlovi** [comment]
@sstok I've added some changes to the query generator itself:

1. `getQuery` now wraps in `query` so you can use it as is (refactored your tests to match)
2. implemented a `registerField` in some manner and refactored the hardcoded mappings in query generator and (your tests)

Can you take a look and point out any glaring errors or do you feel this might be a way to go?

**Comment by dkarlovi** [comment]
@sstok 

I've taken the McDonald's principle and have made [a property notation format proposition](https://github.com/rollerworks/search-elasticsearch/pull/1/files#diff-4a56eb64cd11341f5e66b764d17cadc9R36). If you have any feedback on this or anything else in this PR, I'd really love to hear it. Also need a review regarding normalizers (specifically date, but others too). 

I think this has quite some potential, need to work out the details.

File: src/QueryConditionGenerator.php [hunk 5]
```diff
@@ -23,8 +25,10 @@
 
 final class QueryConditionGenerator
 {
+    private const PROPERTY_ID = '_id';
+
     private $searchCondition;
-    private $fieldSet;
+    // private $fieldSet;
```
Comment (sstok) [inline src/QueryConditionGenerator.php]: This property is used in the code?

File: src/QueryConditionGenerator.php [hunk 6]
```diff
@@ -81,40 +105,79 @@ private function processGroup(ValuesGroup $group): array
...
+            $dateTransformer = new DateTimeToStringTransformer(null, 'UTC', 'Y-m-d');
+
             if ($valuesBag->hasSimpleValues()) {
-                $bool[$includingType][]['terms'] = [$this->mappings[$fieldName]->indexName => array_values($valuesBag->getSimpleValues())];
+                if ($fieldType instanceof DateType) {
+                    // date lookups behave like a range even for an exact value in Elasticsearch
+                    // https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#ranges-on-dates
+                    $values = array_values($valuesBag->getSimpleValues());
+                    foreach ($values as $idx => $value) {
+                        $valuesBag->removeSimpleValue($idx);
```
Comment (sstok) [inline src/QueryConditionGenerator.php]: Note a ValuesBag is not immutable, and the QueryConditionGenerator should not change the ValuesBag.
Better to clone this.

For the future: Usually when a Query generator needs something special is to use TypeExtensions with options like `elastic_search_structure_transformer` which receives the ValuesBag (or ValuesGroup, or both) and allows to change it's structure.

File: src/QueryConditionGenerator.php [hunk 7]
```diff
@@ -81,40 +105,79 @@ private function processGroup(ValuesGroup $group): array
...
 
-                $bool['must_not'][] = [$this->mappings[$fieldName]->indexName => $rangeParams];
+                    $bool['must_not'][]['range'][$propertyName] = $rangeParams;
+                }
+            } else {
+                $ids = [];
+                foreach ($valuesBag->get(Range::class) as $range) {
+                    $ids = array_merge($ids, range($range->getLower(), $range->getUpper()));
+                }
+                if (false === empty($ids)) {
```
Comment (sstok) [inline src/QueryConditionGenerator.php]: No need for the explicit Boolean check, you can simple use `if (f!empty($ids)) {`
Same goes for all other code occurrences.

File: src/QueryConditionGenerator.php [hunk 8]
```diff
@@ -81,40 +105,79 @@ private function processGroup(ValuesGroup $group): array
...
-            foreach ($valuesBag->get(Range::class) as $range) {
-                $rangeParams = [
-                    $range->isLowerInclusive() ? 'lte' : 'lt' => $range->getLower(),
-                    $range->isUpperInclusive() ? 'gte' : 'gt' => $range->getUpper(),
-                ];
+            $isId = self::PROPERTY_ID === $propertyName;
 
-                $bool[$includingType][] = [$this->mappings[$fieldName]->indexName => $rangeParams];
-            }
+            if (false === $isId) {
```
Comment (sstok) [inline src/QueryConditionGenerator.php]: This check needs to be done sooner as simple values can also apply to the id, no?

File: src/QueryConditionGenerator.php (hunk 5 above)
Comment (dkarlovi) [reply src/QueryConditionGenerator.php]: Yep, my bad. :(

File: src/QueryConditionGenerator.php (hunk 6 above)
Comment (dkarlovi) [reply src/QueryConditionGenerator.php]: I'm just working on this and it's annoying as it is, especially since it's only placeholder code.

Comment (sstok) [reply src/QueryConditionGenerator.php]: I did something similar for the Doctrine DBAL integration 😋 `query conversions`.
https://github.com/rollerworks/search-doctrine-dbal/blob/master/src/Query/QueryField.php#L96

For example the https://github.com/rollerworks/search-doctrine-dbal/blob/master/src/Extension/Conversion/AgeDateConversion.php allows to search with both an age and a date.

File: src/ValueConversion.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * (c) Sebastiaan Stok <s.stok@rollerscapes.net>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Rollerworks\Component\Search\Elasticsearch;
+
+/**
+ * Interface ValueConversion.
```
[the remark this answers is not in the document]
Comment (dkarlovi) [reply src/ValueConversion.php]: Thanks. I think I have a good idea how we can push the functionality further. Basically, special cases can go to the per-type extension which should allow, for example, exceptions for dates and IDs while keeping the basic query generator quite simple.

It'll be probably done monday. How do you plan actually implementing the client, if not using Elastica?

File: src/QueryConditionGenerator.php (hunk 8 above)
Comment (dkarlovi) [reply src/QueryConditionGenerator.php]: Yes, I'll be moving the whole thing.

File: src/QueryConditionGenerator.php [hunk 10]
```diff
@@ -239,10 +236,53 @@ private function processPatternMatchers(array $values, string $fieldName, array
...
+    private function convertValue($value, ?ValueConversion $converter)
+    {
+        if (null === $converter) {
+            return $value;
+        }
+
+        return $converter->convertValue($value);
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (dkarlovi) [reply src/QueryConditionGenerator.php]: TBH I'm not a big fan of the selective decision process if a docblock is necessary. It puts burden on the developer / reviewer to figure out. My path usually is to add them everywhere, it doesn't hurt readability, but it does allow for my tooling to work better.

I'll remove it here, obviously. :+1:

**Comment by dkarlovi** [comment]
Added a query conversion concept (which should handle special cases for IDs, dates, etc).

Also, you've inverted less than / greater than in your code and tests, I had a WTF moment there for a bit. :sob: Check the latest changes and feedback if you get the chance.

BTW how do I explain to the system that any date-y field should be converted by my DateConverter? Currently, BirthdayType is broken as it's not handled.

**Comment by dkarlovi** [comment]
I've added support for nested queries (for example, binding property `items[].label` will do a nested query on `items` and do a regular query on `items.label`), will finish compare and matchers tomorrow.

What can we do to get this merged and released tomorrow? I'd like to include a preview version of this in my app ASAP (that's the reason I've done this work in the first place). If you could help me out here, it would be great.

Thanks.

**Comment by sstok** [comment]
Amazing work! This looks good enough to move forwards.

About the Client part, I didn't want to force a specific client implementation to use the user.
For Elastica we can pass the generated query to https://github.com/ruflin/Elastica/blob/master/lib/Elastica/Query.php 👍

For the API-Platform we need a full integration (using Elastica) to also make the mapping of Resources to Indexes/Documents work. But I'm not sure what the best approach is for this.
For limiting access we need to have some Context Listener that will (based on a custom implementation) ensures the the ElasticSearch query limits the results to what's accessible by the user.

**Comment by dkarlovi** [comment]
> Amazing work! This looks good enough to move forwards.

\o/ Could you submit the package to Packagist, it would allow us to move forward on this.

> For the API-Platform we need a full integration (using Elastica) to also make the mapping of Resources to Indexes/Documents work. But I'm not sure what the best approach is for this.

Maybe require FOS Elastica Bundle and compiler-pass the mapping into it?

**Comment by sstok** [comment]
I'm going to merge this one so we can move forward 👍

**Comment by sstok** [comment]
Thank you @dkarlovi

**Comment by dkarlovi** [comment]
Wait, I'm pushing more stuff!

**Comment by sstok** [comment]
No problem ;) I can even give you push access to this repository as you did an amazing job and I don't have always time to review everything in detail 😄

Basically I have a few rules (_I still need to formalize these_):

* Don't use the GitHub UI for making direct commits (except for fixing small typo's) - Use pull requests with [clear commit messages](https://chris.beams.io/posts/git-commit/) and use Git rebase (don't merge) for updating your local branch, be sure to clean-up any WIP commits afterwards (I didn't do this for this pull request as it would be to much work with set-up of the master branch and all).

* Use https://github.com/park-manager/hubkit for merging pull-requests and making releases (no major releases yet!!)

* Don't make a make major release (everything for RollerworksSearch v2 is still in alpha atm). Feel free to make unstable 0.x minor/patch releases (using Hubkit), note that you need to have [git signing](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work) enabled for this.

When in doubt about something feel free to ask me 👍 (I will always do a review afterwards)
Again thank you every much for your amazing work and trust in RollerworksSearch 😸

**Comment by dkarlovi** [comment]
> I can even give you push access to this repository

As said before, I really think you should consider a mono-repo for this project, the overhead of setup required for every single-repo will be quite high. Not that I'm requesting push access for that mono repo. :)

For example, I like to setup [all the infra stuff](https://github.com/dkarlovi/dockerfile-dev) to be run from a Makefile and then can just run `make` before commiting.

> clear commit messages

That's why I'd rather you undo this merge as I've been doing WIP commits which I intended to fix afterward.

**Pull request merged**